// katacoda-cli/lib/commands/courses/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const fs = require("fs");
const inquirer = require("inquirer");
const templates_1 = require("../../helpers/templates");
const validation_service_1 = require("../../helpers/validation-service");
class CoursesCreate extends command_1.Command {
    createPathwayFile(course) {
        templates_1.pathwayTemplate.title = course.title;
        templates_1.pathwayTemplate.description = course.description;
        const data = JSON.stringify(templates_1.pathwayTemplate, null, 2);
        fs.writeFileSync(`${course.url}-pathway.json`, data, "utf8");
    }
    async getCourseDetails() {
        const responses = await inquirer.prompt([
            {
                name: "title",
                message: "Course Title: ",
                validate: validation_service_1.validateTitle
            },
            {
                name: "description",
                message: "Course Description: ",
                validate: validation_service_1.validateDescription
            },
            {
                name: "url",
                message: "friendly-url (This will also be the course's folder name): ",
                validate: validation_service_1.validateFriendlyUrl
            }
        ]);
        return responses;
    }
    async run() {
        try {
            const course = await this.getCourseDetails();
            fs.mkdirSync(course.url);
            this.createPathwayFile(course);
            return this.log("Course created successfully.");
        }
        catch (err) {
            return this.error(`Error: ${err.toString()}`);
        }
    }
}
CoursesCreate.description = "Create a new Katacoda course";
exports.default = CoursesCreate;
