// katacoda-cli/lib/commands/courses/delete.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const inquirer = require("inquirer");
const validation_service_1 = require("../../helpers/validation-service");
const fs = require("fs");
const rimraf = require('rimraf');
const path = require('path');
class CoursesDelete extends command_1.Command {
    async getCourse() {
        return await inquirer.prompt([
            {
                name: 'path',
                message: 'Please enter course path',
                default: process.cwd(),
                validate: validation_service_1.validateCourse
            },
            {
                type: 'confirm',
                name: 'confirmation',
                message: `Are you sure you would like to delete this course?`,
                default: false
            }
        ]);
    }
    async run() {
        try {
            const course = await this.getCourse();
            const friendlyUrl = validation_service_1.getFriendlyUrlFromPath(course.path);
            if (course.confirmation) {
                rimraf.sync(course.path);
                fs.unlinkSync(`${path.join(course.path, '../')}${friendlyUrl}-pathway.json`);
                return this.log('Course succesfully deleted.');
            }
        }
        catch (err) {
            return this.error(`Error: ${err.toString()}`);
        }
    }
}
CoursesDelete.description = 'Delete a Katacoda course';
exports.default = CoursesDelete;
