// katacoda-cli/lib/helpers/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultImages = {
    Other: "other",
    Docker: "docker",
    "Docker Swarm": "swarm",
    Kubernetes: "kubernetes",
    "Kubernetes Cluster": "kubernetes-cluster",
    CoreOS: "coreos",
    Ubuntu: "ubuntu",
    Ansible: "ansible",
    "Node.js v6": "node6",
    Go: "go",
    "C#": "c#",
    ".Net Core": "dotnet",
    Java: "java8",
    Bash: "bash"
};
// This should be given by the API.
// editor.katacoda.com/validation/indexjson/layout
exports.defaultLayouts = {
    Terminal: 'terminal',
    '2x Terminal Split': 'terminal-terminal',
    'Editor + Terminal': 'editor-terminal',
    'Terminal + Editor for the embedded presentation': 'editor-terminal-split',
    'Terminal + Iframe': 'terminal-iframe'
};
exports.defaultDifficultyLevels = [
    'Beginner',
    'Intermediate',
    'Advanced'
];
exports.defaultLayoutsImages = {
    'Docker Swarm': 'terminal-terminal',
    'Kubernetes Cluster': 'terminal-terminal'
};
